draco/util/_fast_tools.pyx

```diff
@@ -80,3 +82,68 @@ def _unpack_product_array_fast(cython.numeric[::1] utv, cython.numeric[:, ::1] m
             pi = (nfeed * (nfeed+1) / 2) - ((nfeed-fi) * (nfeed-fi + 1) / 2) + (fj - fi)
 
             mat[i, j] = utv[pi].conjugate()
+
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _calc_redundancy(float[:, ::1] input_flags, int16_t[:, ::1] prod_map, uint32_t[::1] stack_index,
+                     int nstack, float[:, ::1] redundancy):
+    """Quickly calculate redundancy.
+
+    Parameters
+    ----------
+    input_flags : np.ndarray[input, time]
+        The input flags. Must be zeros or ones.
+    prod_map : np.ndarray[prod, 2]
+        The product map.
+    stack_index : np.ndarray[prod]
+        The stack map.
+    nstack : int
+        Number of stacks.
+    redundancy : np.ndarray[nstack, ntime]
+        Array in which to fill out the redundancy of each stack.
+    """
+    
+    cdef int istack
+    cdef int ia, ib
+    cdef int ii, jj
+    
+    cdef int ninput = input_flags.shape[0]
+    cdef int ntime = input_flags.shape[1]
+
+    cdef int nt
+    
+    # Check the array shapes
+    # Need to construct rshape as redundancy.shape, has extra entries as it's of a memoryview
+    rshape = (redundancy.shape[0], redundancy.shape[1])
+    if rshape != (nstack, ntime):
+        raise RuntimeError("redundancy array shape %s incorrect, expected %s" %
+                           (repr(rshape), repr((nstack, ntime))))
+        
+    # Check that we don't index out of bounds from the prod_map
+    if np.min(prod_map) < 0 or np.max(prod_map) >= ninput:
+        raise RuntimeError("Input index in prod_map out of bounds.")
+
+    with nogil:
```

Comment:
since there is no parallelism, why do you need to release the GIL?

Reply:
It was once a parallel loop, but that didn't end up saving much time. In fact many times it even hurt performance. However, if I can release the GIL I might as well, so I left it in.

Reply:
ok, I guess it's not hurting anything